oak_functions/loader/src/metrics.rs

```diff
@@ -0,0 +1,169 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::logger::Logger;
+use log::Level;
+use rand::{distributions::Open01, thread_rng, Rng};
+use serde::Deserialize;
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{Arc, Mutex},
+};
+
+/// Configuration for differentially-private metrics reporting.
+#[derive(Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct PrivateMetricsConfig {
+    pub epsilon: f64,
+    pub batch_size: usize,
+    pub allowed_labels: Vec<String>,
+}
+
+/// Aggregator for count-based differentially private metrics.
```

Comment:
Could you mention what happens when the threshold is reached? It looks like it logs it out using the standard logger, we may also want some other mechanism in the future. In fact I think having the logger inside this struct is unnecessarily strict, and also makes testing harder. It should be possible to just get the aggregated result (as values ideally instead of a string), and for the caller to log it in whatever way it wants, or, in tests, to make assertions on it. And in general, it is preferable to keep side effects separate from pure computation.

Reply:
I think the API should be more "synchronous". When the `report_events` method is called, the caller should get back either None, or the aggregated values (and internally reset the state). Then the caller decides what to do with it. Also in this way there is no need for a separate testing method, I think.

Reply:
Done. I don't really like all the conditional compilation in there, but could not think of a better way to do it.